proposals/p2006.md

```diff
@@ -0,0 +1,654 @@
+the C++ standard) and indexed access (`*` and `[]`). Because these operators
+specifically model forming an L-value and because the return of the operator
+definition is directly used as the expression, it is necessary to return a
+reference to the already-dereferenced object. Returning a pointer would break
+genericity with builtin pointers and arrays in addition to adding a very
+significant syntactic overhead.
+
+#### Member and subobject accessors
+
+Another common use of references is in returns from member functions to provide
+access to a member or subobject, whether const or mutable. This particular use
+case is worth calling out specially as it has an interesting property: this is
+often not a fully indirect access. Instead, it is often simply selecting a
+particular member, field, or other subobject of the data structure. As a
+consequence, making subsequent access transparent seems especially desirable.
+
+However, it is worth noting that this particular use case is also an especially
+common source of lifetime bugs. A classic and pervasive example can be seen when
+calling such a method on a temporary object. The returned reference is almost
+immediately invalid.
+
+#### Non-null pointers
+
+A common reason for using mutable references outside of what has already been
+described is to represent _non-null pointers_ with enforcement in the type
+system. Because the canonical pointer types in C++ are allowed to be null,
+systems that forbid a null in the type system use references to induce any null
+checks to be as early as possible. This effects a
+"[shift left](https://en.wikipedia.org/wiki/Shift-left_testing)" of handling
+null pointers, both moving the error closer to its cause logically and
+increasing the chance of moving earlier in the development process by making it
+a static property enforced at compile time.
+
+References are imperfectly suited to modeling non-null pointers because they are
+missing many of the fundamental properties of pointers such as being able to
+rebind them, being able to take their address, etc. Also, references cannot be
+safely made `const` in the same places that pointers can because that might
+unintentionally change their semantics by allowing temporaries or extending
+lifetimes.
+
+#### Syntax-free dereference
+
+Beyond serving as a non-null pointer, the other broad use case for references is
+to remove the syntactic overhead of taking an address and dereferencing
+pointers. In other words, they provide a way to have _syntax-free dereferences_.
+Outside of function parameters, removing this distinction may provide a
+genericity benefit, as it allows using the same syntax as would be used with
+non-references. In theory code could simply use pointers everywhere, but this
+would add syntactic overhead compared to local variables and references. For
+immutable code, the syntactic overhead seems unnecessary and unhelpful. However,
```

Comment:
> For immutable code,

Is the code what is immutable here, or is it code that is not performing mutation?

Reply:
Switched to "immutable accesses".